aten/src/ATen/native/cpu/CopyKernel.cpp

```diff
@@ -84,22 +84,20 @@ void copy_kernel(TensorIterator& iter, bool /*non_blocking*/) {
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, dtype, "copy_", [&] {
       using dest_t = scalar_t;
       AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, iter.dtype(1), "copy_", [&] {
-        // Note (@zasdfgbnm):
-        //
-        // The code below can not be simplified as
-        //    cpu_kernel(iter, c10::static_cast_with_inter_type<dest_t, scalar_t>::apply);
-        //
-        // because this would force the compiler to instantiate the inline function and generate a function call in the loop
-        // instead of inlining it, making all the optimizations like vectorization impossible.
-        // You can verify this by looking the the symbols of `libtorch_cpu.so`:
-        //
-        //    readelf -Ws libtorch_cpu.so | grep static_cast_with_inter_type
-        //
-        // If done correctly, the above command should have no output.
-        //
-        // See: https://github.com/pytorch/pytorch/issues/31271
-        cpu_kernel(iter, [](scalar_t src) -> dest_t {
-          return c10::static_cast_with_inter_type<dest_t, scalar_t>::apply(src); });
+        if (iter.has_contiguous_first_dim()) {
+          TORCH_INTERNAL_ASSERT(iter.ninputs() == 1);
+          TORCH_INTERNAL_ASSERT(iter.noutputs() == 1);
+
+          iter.for_each([](char **data, const int64_t *strides, int64_t size) {
+            auto src = reinterpret_cast<const scalar_t*>(data[1]);
+            auto dst = reinterpret_cast<dest_t*>(data[0]);
+            at::vec::convert(src, dst, size);
+          });
```

Comment:
`cpu_kernel_vec` doesn't support outputs with different types than the inputs
https://github.com/pytorch/pytorch/blob/495aa9bc3adf88e1499a042323871dc4650c7c6f/aten/src/ATen/native/cpu/Loops.h#L196-L199

That would be really hard to support because the number of elements in `Vectorized<scalar_t>` depends on the size of `scalar_t`.